lib/TH/generic/THStorage.h

```diff
@@ -28,6 +29,8 @@ typedef struct THStorage
     char flag;
     THAllocator *allocator;
     void *allocatorContext;
+    int type;
```

Comment:
Actually, I still don't understand why this is needed. If you have C code that can handle arbitrary storages, you should know that type of storage you're dealing with, don't you? Maybe you can elaborate a bit more on this?